File: dashboard/src/lib/deploy-events.ts
// Hand-written counterpart to api/models/deploy_events.py.
// The codegen output at dashboard/src/lib/deploy-events.gen.ts (Task A8)
// matches this shape; this file is the import target for the rest of the
// dashboard until we wire the codegen into the typecheck step.

export type DeployEventType = "phase" | "log" | "complete" | "error";

export type DeployPhase =
  | "provisioning"
  | "building"
  | "pushing"
  | "deploying"
  | "health_checking"
  | "registering";

export type DeployJobStatus =
  | "pending"
  | "parsing"
  | "pending_approval"
  | "provisioning"
  | "building"
  | "pushing"
  | "deploying"
  | "health_checking"
  | "registering"
  | "completed"
  | "failed"
  | "timed_out";

export interface DeployEvent {
  type: DeployEventType;
  job_id: string;
  timestamp: string;
  phase: DeployPhase | null;
  step: number | null;
  total: number | null;
  message: string | null;
  level: "info" | "warn" | "error" | null;
  endpoint_url: string | null;
  error_code: string | null;
}

File: api/models/deploy_events.py
"""DeployEvent — wire format for the SSE deploy progress stream (#387)."""

from __future__ import annotations

from datetime import datetime
from enum import StrEnum
from typing import Literal

from pydantic import BaseModel, Field


class DeployJobStatus(StrEnum):
    PENDING = "pending"
    PENDING_APPROVAL = "pending_approval"
    PROVISIONING = "provisioning"
    BUILDING = "building"
    PUSHING = "pushing"
    DEPLOYING = "deploying"
    HEALTH_CHECK = "health_check"
    REGISTERING = "registering"
    COMPLETED = "completed"
    FAILED = "failed"
    TIMED_OUT = "timed_out"


PhaseName = Literal[
    "provisioning", "building", "pushing", "deploying", "health_check", "registering",
]
EventType = Literal["phase", "log", "complete", "error"]


class DeployEvent(BaseModel):
    """One event on the per-job SSE stream. Discriminated by `type`."""

    type: EventType
    job_id: str
    timestamp: datetime
    phase: PhaseName | None = None
    step: int | None = None          # 1-based within phase
    total: int | None = None         # phase total steps
    message: str | None = None
    level: Literal["info", "warn", "error"] | None = None  # only for type="log"
    endpoint_url: str | None = None  # only for type="complete"
    error_code: str | None = None    # only for type="error"

    model_config = {"frozen": True}  # events are immutable once published
